pkg/daemon/daemon.go

```diff
@@ -169,23 +163,39 @@ func (d *Daemon) GetState() (types.DaemonState, error) {
 //  2. READY
 //  3. RUNNING
 func (d *Daemon) WaitUntilState(expected types.DaemonState) error {
-	return retry.Do(func() error {
-		state, err := d.GetState()
-		if err != nil {
-			return errors.Wrapf(err, "wait until daemon is %s", expected)
-		}
+	stateGetter := func() (v any, err error) {
+		err = retry.Do(func() error {
+			if expected == d.State {
+				return nil
+			}
 
-		if state != expected {
-			return errors.Errorf("daemon %s is not %s yet, current state %s",
-				d.ID(), expected, state)
-		}
+			state, err := d.GetState()
+			if err != nil {
+				return errors.Wrapf(err, "wait until daemon is %s", expected)
+			}
+
+			if state != expected {
+				return errors.Errorf("daemon %s is not %s yet, current state %s",
+					d.ID(), expected, state)
+			}
+
+			return nil
+		},
+			retry.Attempts(20), // totally wait for 2 seconds, should be enough
+			retry.LastErrorOnly(true),
+			retry.Delay(100*time.Millisecond),
+		)
+
+		return
+	}
+
+	var g singleflight.Group
```

Comment:
Why new the `singleflight.Group` here, how does it ensures only one is inflight?

Reply:
Good catch. 👍

Reply:
fixed